pkg/oci/build_artifacthub_pkg.go

```diff
@@ -0,0 +1,198 @@
+    sudo ig run {{ image }}
+    ` + "```" + `
+#changes:
+#    - kind: added
+#      description: Initial release
+provider:
+    name: Inspektor Gadget
+`
+
+func createOrUpdateArtifactHubPkg(ctx context.Context, opts *BuildGadgetImageOpts) error {
+	// load metadata file
+	metadataFile, err := os.Open(opts.MetadataPath)
+	if err != nil {
+		return fmt.Errorf("opening metadata file: %w", err)
+	}
+	defer metadataFile.Close()
+
+	metadata := &types.GadgetMetadata{}
+	if err := yaml.NewDecoder(metadataFile).Decode(metadata); err != nil {
+		return fmt.Errorf("decoding metadata file: %w", err)
+	}
+
+	_, statErr := os.Stat(opts.ArtifactHubPkgPath)
+	update := statErr == nil
+
+	var ArtifactHubPkgBytes []byte
+	if update {
+		// load artifacthub-pkg.yml
+		ArtifactHubPkgBytes, err = os.ReadFile(opts.ArtifactHubPkgPath)
+		if err != nil {
+			return fmt.Errorf("reading artifact hub pkg file: %w", err)
+		}
+
+		log.Debugf("Artifact hub pkg file found, updating it")
+	} else {
+		log.Debug("Artifact hub pkg file not found, generating it")
+
+		imageFunc := func() string {
+			// Try to guess the image name from the source URL
+			githubUsername := "username"
+			if strings.HasPrefix(metadata.SourceURL, "https://github.com/") {
+				parts := strings.Split(metadata.SourceURL, "/")
+				if len(parts) >= 4 {
+					githubUsername = parts[3]
+				}
+			}
+			dir := filepath.Dir(opts.MetadataPath)
+			if dir == "." {
+				dir, _ = os.Getwd()
+			}
+			dir = filepath.Base(dir)
+
+			dirParts := strings.Split(dir, "_")
+			imagePrefix := ""
+			if len(dirParts) > 1 {
+				imagePrefix = dirParts[0] + "_"
+			}
+			shortImageName := strings.Replace(metadata.Name, " ", "_", -1)
+			if strings.HasPrefix(shortImageName, imagePrefix) {
+				imagePrefix = ""
+			}
+			return fmt.Sprintf("ghcr.io/%s/gadget/%s%s:latest", githubUsername, imagePrefix, shortImageName)
+		}
+
+		funcMap := template.FuncMap{
+			"image": imageFunc,
+		}
+
+		t, err := template.New("artifacthub-pkg.yml").Funcs(funcMap).Parse(ArtifactHubPkgTemplate)
+		if err != nil {
+			return fmt.Errorf("parsing artifact hub pkg template: %w", err)
+		}
+		var doc bytes.Buffer
+		t.Execute(&doc, metadata)
+
+		ArtifactHubPkgBytes = []byte(doc.String())
+	}
+
+	var artifactHubPkg yaml.Node
+	err = yaml.Unmarshal(ArtifactHubPkgBytes, &artifactHubPkg)
+	if err != nil {
+		return fmt.Errorf("unmarshalling artifact hub pkg file: %w\n%s", err, string(ArtifactHubPkgBytes))
+	}
+
+	if artifactHubPkg.Kind != yaml.DocumentNode {
+		return fmt.Errorf("artifact hub pkg file is not a document")
+	}
+	if len(artifactHubPkg.Content) == 0 {
+		return fmt.Errorf("artifact hub pkg file is empty")
+	}
+	if artifactHubPkg.Content[0].Kind != yaml.MappingNode {
+		return fmt.Errorf("artifact hub pkg file is not a mapping: %v", artifactHubPkg.Content[0].Kind)
+	}
+	if len(artifactHubPkg.Content[0].Content)%2 != 0 {
+		return fmt.Errorf("artifact hub pkg file is not a mapping: len=%d", len(artifactHubPkg.Content[0].Content))
```

Comment:
The check if the pkg file is a mapping was done right before this check. The error message still states it is not a mapping. (I have no idea what a mapping is or what `%2 != 0` is for. Maybe my comment doesn't make sense)

Reply:
A sequence would be something like this in json:
```json
[
  "entry1",
  "entry2"
]
```
A mapping would be something like this in json:
```json
{
  "key1": "value1",
  "key2": "value2"
}
```
The top-level data structure in `artifacthub-pkg.yml` is a mapping.

The yaml parsing library would store the entries of the mapping in the variable:
```golang
// pseudo Golang code
var Content []*Node = []*Node{&Node{"key1"}, &Node{"value1"}, &Node{"key2"}, &Node{"value2"}}
```
So a valid mapping always has an even number of nodes because there is one value for each key.

I'll improve the message